tests/unit/Alignment.cpp

```diff
@@ -0,0 +1,283 @@
+                 Expected(model::ABI::SystemZ_s390x, 0),
+                 Expected(model::ABI::SystemV_x86, 0));
+  ConstantTest Float4Test{ makePrimitive(Primitive::Float, 4, *Binary) };
+  Float4Test.run(Expected(model::ABI::AAPCS64, 4),
+                 Expected(model::ABI::AAPCS, 4),
+                 Expected(model::ABI::SystemZ_s390x, 4),
+                 Expected(model::ABI::SystemV_x86, 4));
+  ConstantTest Float8Test{ makePrimitive(Primitive::Float, 8, *Binary) };
+  Float8Test.run(Expected(model::ABI::AAPCS64, 8),
+                 Expected(model::ABI::AAPCS, 8),
+                 Expected(model::ABI::SystemZ_s390x, 8),
+                 Expected(model::ABI::SystemV_x86, 4));
+  ConstantTest Float16Test{ makePrimitive(Primitive::Float, 16, *Binary) };
+  Float16Test.run(Expected(model::ABI::AAPCS64, 16),
+                  Expected(model::ABI::AAPCS, 0),
+                  Expected(model::ABI::SystemZ_s390x, 8),
+                  Expected(model::ABI::SystemV_x86, 16));
+}
+
+constexpr std::array TestedABIs{ model::ABI::AAPCS64,
+                                 model::ABI::AAPCS,
+                                 model::ABI::SystemZ_s390x,
+                                 model::ABI::SystemV_x86 };
+
+BOOST_AUTO_TEST_CASE(RemainingPrimitiveTypes) {
+  TupleTree<model::Binary> Binary;
+
+  constexpr std::array<model::PrimitiveTypeKind::Values, 5> RemainingTypes{
+    model::PrimitiveTypeKind::Unsigned,
+    model::PrimitiveTypeKind::Signed,
+    model::PrimitiveTypeKind::Number,
+    model::PrimitiveTypeKind::PointerOrNumber
+  };
+  for (const auto &PrimitiveKind : RemainingTypes) {
+    for (std::uint64_t Size = 1; Size <= 16; Size *= 2) {
+      ComparisonTest GenericTest{ TestedABIs };
+      GenericTest.run(makePrimitive(Primitive::Generic, Size, *Binary),
+                      makePrimitive(PrimitiveKind, Size, *Binary));
+    }
+  }
+}
+
+static model::UnionField
+makeUnionField(std::uint64_t Index, model::QualifiedType Type) {
+  model::UnionField Result;
+  Result.Index() = Index;
+  Result.Type() = Type;
+  return Result;
+}
+
+BOOST_AUTO_TEST_CASE(UnionTypes) {
+  TupleTree<model::Binary> Binary;
+
+  using QT = model::QualifiedType;
+  QT Int16 = makePrimitive(model::PrimitiveTypeKind::Signed, 2, *Binary);
+  QT Int32 = makePrimitive(model::PrimitiveTypeKind::Signed, 4, *Binary);
+  QT Int64 = makePrimitive(model::PrimitiveTypeKind::Signed, 8, *Binary);
+  QT Float = makePrimitive(model::PrimitiveTypeKind::Float, 4, *Binary);
+  QT Double = makePrimitive(model::PrimitiveTypeKind::Float, 8, *Binary);
+  QT LongDouble = makePrimitive(model::PrimitiveTypeKind::Float, 16, *Binary);
+  QT WeirdLD = makePrimitive(model::PrimitiveTypeKind::Float, 12, *Binary);
+
+  auto [SimpleU, SimplePath] = Binary->makeType<model::UnionType>();
+  SimpleU.Fields().emplace(makeUnionField(0, Int32));
+  SimpleU.Fields().emplace(makeUnionField(1, Int64));
+  ComparisonTest{ TestedABIs }.run(Int64, QT(SimplePath, {}));
```

Comment:
Please make a `testUnionTypes` function that takes an ABI as argument and have an outer loop testing all the ABIs we're interested in.

The current testing approach is a bit counterintuitive, I guess you were trying to optimize performance and avoid creating types for every ABI, but unit tests literally take a blink of an eye, so let's prefer readability.

Reply:
Actually, now that I think of it, there's no real need to create so many `ComparisonTest` objects, it can be simplified to
```cpp
ComparisonTest Tester{ TestedABIs };
```
at the beginning of the test, and then for each function
```cpp
Tester.compare_alignment(FirstType, SecondType); 
```
IMO the biggest problem with readability here is actually the `run` name, which doesn't reflect what it is doing.

Also, maybe I should have actually gone and made the union creation tester, so that it looks like

```cpp
auto SimpleUnion = makeUnion(makeUnionField(0, Int32), makeUnionField(1, Int64));
Tester.compare_alignment(SimpleUnion, Int64);
```

Thoughts? Opinions?

Reply:
The problem is not much about how many `ComparisonTest` objects you create but on the fact that the is basically:

```c++
for (type : types_to_test)
  for (abi : abis)
    test(abi, type);
```

While I'd like

```c++
for (abi : abis)
  for (type : types_to_test)
    test(abi, type);
```
 
And basically the inner loop is just a function that takes the ABI as an argument.

> IMO the biggest problem with readability here is actually the run name, which doesn't reflect what it is doing.

AFAIU `run` means "run on all ABIs", but as said above, I'd like to invert how we test and have a bunch of simple assertions. `run` becomes unnecessary then.

> maybe I should have actually gone and made the union creation tester

My comment was not specific to `union`s, I'd like the same for `struct`s.
Anyway, if you think that's beneficial go for it.

Reply:
Yeah, I guess I was too focused on trying to preserve the similarities with the other test type (comparison of primitive alignments to hard-coded values) where we cannot move ABI out, since the test values depend on them heavily.

You suggestion is indeed easier to read, sorry for not realizing it sooner.

> My comment was not specific to `union`s, I'd like the same for `struct`s.

Yes-yes, the tests on the two are practically the same anyway, so when I'm talking about the changes to one type, to the other one is also implied. Sorry for not being clearer on that.

Reply:
What about now?

Reply:
I still see:

```c++
for (type : types_to_test)
  for (abi : abis)
    test(abi, type);
```

I was suggesting to move the inner `for` out.

Reply:
I didn't push. It's the same state you looked like the previous time 😅.

Sorry, please take a look now.

Reply:
And while at it, I also added an extra tests for pointers, arrays and such I forgot about initially, please take an extra look at the last `BOOST_AUTO_TEST_CASE` as well.

Reply:
Added some comments, it mostly looks good now.